src/algorithms/all_paths.c

```diff
@@ -26,6 +26,38 @@ static bool _AllPathsCtx_LevelNotEmpty(const AllPathsCtx *ctx, uint level) {
 	return (level < array_len(ctx->levels) && array_len(ctx->levels[level]) > 0);
 }
 
+// Traverse from the frontier node in the specified direction and add all encountered nodes and edges.
+static void _addNeighbors(AllPathsCtx *ctx, LevelConnection *frontier, uint32_t depth,
+						  GRAPH_EDGE_DIR dir) {
+	EntityID frontierId = INVALID_ENTITY_ID;
+	if(depth > 1) frontierId = ENTITY_GET_ID(&frontier->edge);
+
+	// Get frontier neighbors.
+	for(int i = 0; i < ctx->relationCount; i++) {
+		Graph_GetNodeEdges(ctx->g, &frontier->node, dir, ctx->relationIDs[i], &ctx->neighbors);
+	}
+
+	// Add unvisited neighbors to next level.
+	uint32_t neighborsCount = array_len(ctx->neighbors);
+	for(uint32_t i = 0; i < neighborsCount; i++) {
+		// Don't follow the frontier edge again.
+		if(frontierId == ENTITY_GET_ID(ctx->neighbors + i)) continue;
+		// Set the neighbor by following the edge in the correct directoin.
+		Node neighbor;
+		switch(dir) {
+		case GRAPH_EDGE_DIR_OUTGOING:
+			Graph_GetNode(ctx->g, Edge_GetDestNodeID(ctx->neighbors + i), &neighbor);
+			break;
+		default:
+			Graph_GetNode(ctx->g, Edge_GetSrcNodeID(ctx->neighbors + i), &neighbor);
+			break;
+		}
```

Comment:
Looking at the code which calls this function I think `dir` can only be:
1. `GRAPH_EDGE_DIR_OUTGOING`
2. `GRAPH_EDGE_DIR_INCOMING`

If that's so, please assert on `GRAPH_EDGE_DIR_BOTH`